piksi_tools/console/observation_view.py

```diff
@@ -198,80 +126,88 @@ def obs_packed_callback(self, sbp_msg, **metadata):
       prn = o.sid.sat
       flags = 0
       msdopp = 0
+
       # Old PRN values had to add one for GPS
-      if (code_is_gps(o.sid.code) 
-            and sbp_msg.msg_type in [SBP_MSG_OBS_DEP_B, SBP_MSG_OBS_DEP_C, SBP_MSG_OBS_DEP_A]):
+      if (code_is_gps(o.sid.code)
+            and sbp_msg.msg_type in [SBP_MSG_OBS_DEP_A, SBP_MSG_OBS_DEP_B, SBP_MSG_OBS_DEP_C]):
         prn += 1
+
       prn = '{} ({})'.format(prn, code_to_str(o.sid.code))
+
       # DEP_B and DEP_A obs had different pseudorange scaling
-      if sbp_msg.msg_type in [SBP_MSG_OBS_DEP_B, SBP_MSG_OBS_DEP_A] :
+      if sbp_msg.msg_type in [SBP_MSG_OBS_DEP_A, SBP_MSG_OBS_DEP_B] :
         divisor = 1e2
       else:
         divisor = 5e1
-      if sbp_msg.msg_type not in [SBP_MSG_OBS_DEP_B, SBP_MSG_OBS_DEP_C, SBP_MSG_OBS_DEP_A]:
+
+      if sbp_msg.msg_type not in [SBP_MSG_OBS_DEP_A, SBP_MSG_OBS_DEP_B, SBP_MSG_OBS_DEP_C]:
         flags = o.flags
         msdopp = float(o.D.i) + float(o.D.f) / (1 << 8)
         self.gps_tow += sbp_msg.header.t.ns * 1e-9
-      
-      # compute time difference of carrier phase for display, only if carrier phase is valid
+
       try:
         ocp = self.old_obs[prn][1]
       except:
         ocp = 0
+
       cp = float(o.L.i) + float(o.L.f) / (1 << 8)
+
+      # compute time difference of carrier phase for display, only if carrier phase is valid
       if ocp != 0 and (flags & 0x3) == 0x3:
-        cf = (cp - ocp) / float(self.gps_tow - self.old_tow)
+        # Doppler per RINEX has opposite sign direction to carrier phase
+        cf = (ocp - cp) / float(self.gps_tow - self.old_tow)
+
+        # Older messages had flipped sign carrier phase values
+        if sbp_msg.msg_type in [SBP_MSG_OBS_DEP_A, SBP_MSG_OBS_DEP_B]:
+          cf = -cf
       else:
         cf = 0
-      self.obs[prn] = (float(o.P) / divisor,
-                       float(o.L.i) + float(o.L.f) / (1 << 8),
-                       float(o.cn0) / 4,
-                       msdopp, cf, o.lock, "0x{:04X}".format(flags))
+
+      flags_string = "0x{:04X} = ".format(flags)
+
+      # Add some indicators to help understand the flags values
+      # Bit 0 is Pseudorange valid
+      if (flags & 0x01):
+        flags_string += "PR "
+      # Bit 1 is Pseudorange valid
+      if (flags & 0x02):
+        flags_string += "CP "
+      # Bit 2 is Pseudorange valid
+      if (flags & 0x04):
+        flags_string += "1/2C "
+      # Bit 3 is Pseudorange valid
+      if (flags & 0x08):
+        flags_string += "MD "
+
+      pr = float(o.P) / divisor
+      cp = float(o.L.i) + float(o.L.f) / (1 << 8)
+      cn0 = float(o.cn0) / 4
+
+      # Sets invalid values to zero
+      # TODO(Leith): it would be better if the cell was blank when invalid
+      if sbp_msg.msg_type not in [SBP_MSG_OBS_DEP_A, SBP_MSG_OBS_DEP_B, SBP_MSG_OBS_DEP_C]:
```

Comment:
If we want to hide invalid values, I think we have to change every column to a string and use the "EMPTY_STR" global when no data exists.  That is currently set to be "--".